### PR TITLE
messagebus initial implementation

### PR DESCRIPTION
**Pull request by fadhillahentino**
Features added :
 - messagebus interface
 - kafka publish async, sync and publish with response
 - kafka register topic and callback (subscribe message) 
 please help me to review @thoriqsatriya @banukusuma @rjjatson @desniaak @purbopanambang @rafi-isakh

### CODE REVIEW

**Comment by anggorodewanto** [comment]
Write README documenting the library's usage please

**Comment by fadhillahentino** [comment]
> Write README documenting the library's usage please

**Comment by fadhillahentino** [comment]
> Write README documenting the library's usage please

Sure, because we have another priority
I delayed writing the README and will update it ASAP. thanks

File: kafka.go [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+
+func listenAsyncError(producer sarama.AsyncProducer) {
+	for err := range producer.Errors() {
+		logrus.Error("Unable to publish message using async producer to kafka : ", err)
+	}
+}
+
+// SetTimeout set listening timeout for publish with response
+func (client *KafkaClient) SetTimeout(timeout int) {
+	if timeout < 10000 {
```
Comment (thoriqsatriya) [inline kafka.go]: don't you think it's better to put this into a const?

File: kafka.go [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+func createTopic(broker *sarama.Broker, topicName string) error {
+	topicDetail := &sarama.TopicDetail{}
+	topicDetail.NumPartitions = int32(1)
+	topicDetail.ReplicationFactor = int16(1)
+	topicDetail.ConfigEntries = make(map[string]*string)
+
+	topicDetails := make(map[string]*sarama.TopicDetail)
+	topicDetails[topicName] = topicDetail
+	request := sarama.CreateTopicsRequest{
+		Timeout:      time.Second * 15,
```
Comment (thoriqsatriya) [inline kafka.go:312]: also this one, better to put in a constant

File: kafka.go [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+	}
+
+	_, err := broker.CreateTopics(&request)
+	return err
+}
+
+// deleteTopic delete kafka topic
+func deleteTopic(broker *sarama.Broker, topicName string) error {
+	request := sarama.DeleteTopicsRequest{
+		Timeout: time.Second * 15,
```
Comment (thoriqsatriya) [inline kafka.go]: also this one, better to put in a constant

File: kafka.go [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+	receivedMessage := unmarshal(consumerMessage)
+	runCallback(receivedMessage, consumerMessage)
+}
+
+// processResponseMessage process a message from kafka for publish with responses function
+func processResponseMessage(consumerMessage *sarama.ConsumerMessage, messageID string) bool {
+	receivedMessage := unmarshal(consumerMessage)
+
+	// check the request and response message ID should be equal
+	if messageID != receivedMessage.MessageId {
```
Comment (thoriqsatriya) [inline kafka.go:391]: is it possible that the receivedMessage is nil?

File: kafka.go [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+}
+
+// NewKafkaClient create a new instance of KafkaClient
+func NewKafkaClient(brokerList []string) (*KafkaClient, error) {
+	config := sarama.NewConfig()
+	config.Version = sarama.V2_1_0_0
+
+	// currently only support 1 message broker
+	broker := sarama.NewBroker(brokerList[0])
+	_ = broker.Open(config)
```
Comment (thoriqsatriya) [inline kafka.go]: is this an error that you silent? or something else?

File: kafka.go (hunk 2 above)
Comment (fadhillahentino) [reply kafka.go:312]: sure, will update it
Thanks

File: messagebus.go [hunk 6]
```diff
@@ -0,0 +1,126 @@
+// Copyright (c) 2019 AccelByte Inc. All Rights Reserved.
+// This is licensed software from AccelByte Inc, for limitations
+// and restrictions contact your company contract manager.
+
+package messagebus
+
+const (
+	GDPR_DeletionAccount string = "GDPR:deletionAccount"
```
Comment (desniaak) [inline messagebus.go]: what is this used for?

File: kafka.go (hunk 4 above)
Comment (fadhillahentino) [reply kafka.go:391]: I think its not possible because if unmarshal function get an error, it will return empty struct

File: kafka.go (hunk 5 above)
Comment (fadhillahentino) [reply kafka.go]: you're right, I will update it
thanks

File: messagebus.go (hunk 6 above)
Comment (fadhillahentino) [reply messagebus.go]: this is a topic name, we registered the topic name as a public constant so the client doesn't need to manually set the topic

File: kafka.go (hunk 3 above)
Comment (fadhillahentino) [reply kafka.go]: sure, I will update it
thanks

File: kafka.go (hunk 1 above)
Comment (fadhillahentino) [reply kafka.go]: sure, I will update it
thanks

**Pull request merged**